gpu-simulator/trace-driven/trace_driven.cc

```diff
@@ -582,7 +584,20 @@ const warp_inst_t *trace_shader_core_ctx::get_next_inst(unsigned warp_id,
   // read the inst from the traces
   trace_shd_warp_t *m_trace_warp =
       static_cast<trace_shd_warp_t *>(m_warp[warp_id]);
-  return m_trace_warp->get_next_trace_inst();
+  const trace_warp_inst_t * ret =  m_trace_warp->get_next_trace_inst();
+  if (m_trace_warp->trace_done() ) {
+    if (!m_warp[warp_id]->inst_in_pipeline() &&
```

Comment:
So we have to test for all these conditions here and in the shader's fetch stage?
What is happening to the front-end when the trace is done, but these other conditions are not met? Are we just returning NULL from this function? And the performance model treats it almost like we are waiting for the iCache to resolve?

Reply:
ok makes sense.